driver_test.go

```diff
@@ -12,26 +12,267 @@
+		}
+		return nil
+	})
+
+	return err
+}
+
+func TestQueryContext(t *testing.T) {
+
+	// Set up test table.
+	conn, err := NewConnector()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer conn.Close()
+
+	err = executeDdlApi(conn, []string{`CREATE TABLE TestQueryContext (
+		A   STRING(1024),
+		B  STRING(1024),
+		C   STRING(1024)
+	)	 PRIMARY KEY (A)`})
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = ExecuteDMLClientLib([]string{`INSERT INTO TestQueryContext (A, B, C) 
+		VALUES ("a1", "b1", "c1"), ("a2", "b2", "c2") , ("a3", "b3", "c3") `})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Open db.
+	ctx := context.Background()
+	db, err := sql.Open("spanner", dsn)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer db.Close()
+
+	type testQueryContextRow struct {
+		A, B, C string
+	}
+
+	tests := []struct {
+		name           string
+		input          string
+		want           []testQueryContextRow
+		wantErrorQuery bool
+		wantErrorScan  bool
+		wantErrorClose bool
+	}{
+		{name: "empty query",
+			input: "", want: []testQueryContextRow{}},
+		{name: "suntax error",
+			input: "SELECT SELECT * FROM TestQueryContext", want: []testQueryContextRow{}},
+		{name: "return nothing",
+			input: "SELECT * FROM TestQueryContext WHERE A = \"hihihi\"", want: []testQueryContextRow{}},
+		{name: "select one tuple",
+			input: "SELECT * FROM TestQueryContext WHERE A = \"a1\"",
+			want: []testQueryContextRow{
+				{A: "a1", B: "b1", C: "c1"},
+			}},
+		{name: "select subset of tuples",
+			input: "SELECT * FROM TestQueryContext WHERE A = \"a1\" OR A = \"a2\"",
+			want: []testQueryContextRow{
+				{A: "a1", B: "b1", C: "c1"},
+				{A: "a2", B: "b2", C: "c2"},
+			}},
+		{name: "select subset of tuples with !=",
+			input: "SELECT * FROM TestQueryContext WHERE A != \"a3\"",
+			want: []testQueryContextRow{
+				{A: "a1", B: "b1", C: "c1"},
+				{A: "a2", B: "b2", C: "c2"},
+			}},
+		{name: "select entire table",
+			input: "SELECT * FROM TestQueryContext ORDER BY A",
+			want: []testQueryContextRow{
+				{A: "a1", B: "b1", C: "c1"},
+				{A: "a2", B: "b2", C: "c2"},
+				{A: "a3", B: "b3", C: "c3"},
+			}},
+		{name: "query non existant table",
+			input: "SELECT * FROM TestQueryContexta", want: []testQueryContextRow{}},
+	}
+
+	// Run tests
+	for _, tc := range tests {
+
+		rows, err := db.QueryContext(ctx, tc.input)
+		if (err != nil) && (!tc.wantErrorQuery) {
+			t.Errorf("%s: unexpected query error: %v", tc.name, err)
+		}
+		if (err == nil) && (tc.wantErrorQuery) {
+			t.Errorf("%s: expected query error but error was %v", tc.name, err)
+		}
+
+		got := []testQueryContextRow{}
+		for rows.Next() {
+			var curr testQueryContextRow
+			err := rows.Scan(&curr.A, &curr.B, &curr.C)
+			if (err != nil) && (!tc.wantErrorScan) {
+				t.Errorf("%s: unexpected query error: %v", tc.name, err)
+			}
+			if (err == nil) && (tc.wantErrorScan) {
+				t.Errorf("%s: expected query error but error was %v", tc.name, err)
+			}
+
+			got = append(got, curr)
+		}
+		rerr := rows.Close()
```

Comment:
Looking at [this issue](https://github.com/golang/go/issues/24329), it seems that the preferred option is to check `rows.Err()` rather than `Rows.Close()`.

Reply:
Also, you can reuse the `err` variable here rather than redeclaring another one :+1:

Reply:
changed